pkg/gardenlet/controller/bastion/bastion_reconciler.go

```diff
@@ -0,0 +1,256 @@
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get seed client: %w", err)
+	}
+
+	// get Shoot for the bastion
+	shoot := gardencorev1beta1.Shoot{}
+	shootKey := kutil.Key(bastion.Namespace, bastion.Spec.ShootRef.Name)
+	if err := gardenClient.Client().Get(ctx, kutil.Key(bastion.Namespace, bastion.Spec.ShootRef.Name), &shoot); err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not get shoot %v: %v", shootKey, err)
+	}
+
+	if bastion.DeletionTimestamp != nil {
+		err = r.cleanupBastion(ctx, logger, gardenClient.Client(), seedClient.DirectClient(), bastion, &shoot)
+	} else {
+		err = r.reconcileBastion(ctx, logger, gardenClient.Client(), seedClient.DirectClient(), bastion, &shoot)
+	}
+
+	if cause := extensionscontroller.ReconcileErrCause(err); cause != nil {
+		logger.Errorf("Reconciling failed: %v", cause)
+	}
+
+	return extensionscontroller.ReconcileErr(err)
+}
+
+func (r *reconciler) reconcileBastion(
+	ctx context.Context,
+	logger logrus.FieldLogger,
+	gardenClient client.Client,
+	seedClient client.Client,
+	bastion *operationsv1alpha1.Bastion,
+	shoot *gardencorev1beta1.Shoot,
+) error {
+	logger.Debug("Reconciling")
+
+	// ensure finalizer is set
+	if !controllerutil.ContainsFinalizer(bastion, finalizerName) {
+		if err := controllerutils.PatchAddFinalizers(ctx, gardenClient, bastion, finalizerName); err != nil {
+			return fmt.Errorf("failed ensure %q finalizer on bastion: %v", finalizerName, err)
+		}
+		// the patch above already triggers a reconcile, so we can stop here
+		return nil
+	}
+
+	// prepare extension resource
+	extBastion := newBastionExtension(bastion, shoot)
+	extensionsIngress := make([]extensionsv1alpha1.BastionIngressPolicy, len(bastion.Spec.Ingress))
+	for i, ingress := range bastion.Spec.Ingress {
+		extensionsIngress[i] = extensionsv1alpha1.BastionIngressPolicy{
+			IPBlock: ingress.IPBlock,
+		}
+	}
+
+	// create or update the bastion in the seed cluster
+	if _, err := controllerutil.CreateOrUpdate(ctx, seedClient, extBastion, func() error {
+		extBastion.Spec.UserData = createUserData(bastion)
+		extBastion.Spec.Ingress = extensionsIngress
+		extBastion.Spec.Type = *bastion.Spec.ProviderType
+
+		return nil
+	}); err != nil {
+		patchReadyCondition(ctx, gardenClient, bastion, gardencorev1alpha1.ConditionFalse, "FailedReconciling", err.Error())
+
+		return fmt.Errorf("failed ensure bastion extension resource: %v", err)
+	}
+
+	// CreateOrUpdate left the extBastion variable updated, so we can conveniently
+	// check the bastion's status in the seed without having to fetch it again.
+	if extBastion.Status.LastOperation == nil || extBastion.Status.LastOperation.State != gardencorev1beta1.LastOperationStateSucceeded {
+		patchReadyCondition(ctx, gardenClient, bastion, gardencorev1alpha1.ConditionFalse, "Reconciling", "Reconciling has not finished yet.")
+
+		return &ctrlerror.RequeueAfterError{
+			RequeueAfter: 1 * time.Second,
```

Comment:
It could. My thought was that during this reconciliation, some users sits in front of their machine and is impatiently waiting for the SSH connection to come up, and I wanted to serve them ASAP.